mitmproxy/optmanager.py

```diff
@@ -1,37 +1,84 @@
 import contextlib
 import blinker
 import pprint
-import inspect
 import copy
 import functools
 import weakref
 import os
+import typing
+import textwrap
 
 import ruamel.yaml
 
 from mitmproxy import exceptions
 from mitmproxy.utils import typecheck
 
-
 """
     The base implementation for Options.
 """
 
+unset = object()
+
+
+class _Option:
+    __slots__ = ("name", "typespec", "value", "_default", "choices", "help")
+
+    def __init__(
+        self,
+        name: str,
+        default: typing.Any,
+        typespec: typing.Type,
+        help: str,
+        choices: typing.Optional[typing.Sequence[str]]
+    ) -> None:
+        typecheck.check_type(name, default, typespec)
+        self.name = name
+        self._default = default
+        self.typespec = typespec
+        self.value = unset
+        self.help = help
+        self.choices = choices
+
+    def __repr__(self):
+        return "{value} [{type}]".format(value=self.current(), type=self.typespec)
+
+    @property
+    def default(self):
+        return copy.deepcopy(self._default)
```

Comment:
Would we ever have mutable defaults? Right now we only have sequences, for which we could use tuples. We don't have dicts yet, but we could use `types.MappingProxyType` as a "frozendict". This would even be a bit nicer as that you cannot accidentally call `.append()` without noticing that it's a no-op.